File: python/board.py
# Creates an instance of Board.
#
# @constructor
# @this {Board}
# @param {Game} game The main-game object.
# @param {array} field The field containing our situation.
# @param {number} player The current player.

import copy


class Board:

    def __init__(self, game, field, player):
        self.game = game
        self.field = field
        self.player = player

    # Determines if situation is finished.
    #
    # @param {number} depth
    # @param {number} score
    # @return {boolean}

    def is_finished(self, depth, score):
        if (depth == 0) | (score == self.game.score) | (score == -self.game.score) | (self.is_full is True):
            return True
        return False

    # Place in current board.
    #
    # @param {number} column
    # @return {boolean}
    def place(self, column):
        # Check if column valid
        # 1. not empty 2. not exceeding the board size
        if (self.field[0][column] is None) & (column >= 0) & (column < self.game.columns):
            i = (self.game.rows - 1)
            # Bottom to top
            while i >= 0:
                if self.field[i][column] is None:
                    # Set current player coin
                    self.field[i][column] = self.player
                    break
                    # Break from loop after inserting
                i -= 1

            self.player = self.game.switch_round(self.player)
            return True
        else:
            return False

    # Return a score for various positions (either horizontal, vertical or diagonal by moving through our board).
    #
    # @param {number} row
    # @param {number} column
    # @param {number} delta_y
    # @param {number} delta_x
    # @return {number}

    def score_position(self, row, column, delta_y, delta_x):
        human_points = 0
        computer_points = 0

        # Save winning positions to arrays for later usage
        self.game.winning_array_human = []
        self.game.winning_array_cpu = []

        # Determine score through amount of available chips
        i = 0
        while i < 4:
            if self.field[row][column] == 0:
                self.game.winning_array_human.append([row, column])
                human_points += 1
                # Add for each human chip
            elif self.field[row][column] == 1:
                self.game.winning_array_cpu.append([row, column])
                computer_points += 1
                # Add for each computer chip
            # Moving through our board
            row += delta_y
            column += delta_x
            i += 1

        # Marking winning/returning score
        if human_points == 4:
            self.game.winning_array = self.game.winning_array_human
            # Computer won (100000)
            return -self.game.score
        elif computer_points == 4:
            self.game.winning_array = self.game.winning_array_cpu
            # Human won (-100000)
            return self.game.score
        else:
            # Return normal points
            return computer_points

    # Returns the overall score for our board.
    #
    # @return {number}

    def score(self):

        vertical_points = 0
        horizontal_points = 0
        diagonal_points1 = 0
        diagonal_points2 = 0

        # Board-size: 7x6 (height x width)
        # Array indices begin with 0
        # => e.g. height: 0, 1, 2, 3, 4, 5

        # Vertical points
        # Check each column for vertical score
        #
        # Possible situations
        #  0  1  2  3  4  5  6
        # [x][ ][ ][ ][ ][ ][ ] 0
        # [x][x][ ][ ][ ][ ][ ] 1
        # [x][x][x][ ][ ][ ][ ] 2
        # [x][x][x][ ][ ][ ][ ] 3
        # [ ][x][x][ ][ ][ ][ ] 4
        # [ ][ ][x][ ][ ][ ][ ] 5

        row = 0
        while row < self.game.rows - 3:
            # Für jede Column überprüfen
            column = 0
            while column < self.game.columns:
                # Die Column bewerten und zu den Punkten hinzufügen
                score = self.score_position(row, column, 1, 0)
                if score == self.game.score:
                    return self.game.score
                if score == -self.game.score:
                    return -self.game.score
                vertical_points += score
                column += 1
            row += 1

        # Horizontal points
        # Check each row's score
        #
        # Possible situations
        #  0  1  2  3  4  5  6
        # [x][x][x][x][ ][ ][ ] 0
        # [ ][x][x][x][x][ ][ ] 1
        # [ ][ ][x][x][x][x][ ] 2
        # [ ][ ][ ][x][x][x][x] 3
        # [ ][ ][ ][ ][ ][ ][ ] 4
        # [ ][ ][ ][ ][ ][ ][ ] 5

        row = 0
        while row < self.game.rows:
            column = 0
            while column < self.game.columns - 3:
                score = self.score_position(row, column, 0, 1)
                if score == self.game.score:
                    return self.game.score
                if score == -self.game.score:
                    return -self.game.score
                horizontal_points += score
                column += 1
            row += 1

        # Diagonal points 1 (left-bottom)
        #
        # Possible situation
        #  0  1  2  3  4  5  6
        # [x][ ][ ][ ][ ][ ][ ] 0
        # [ ][x][ ][ ][ ][ ][ ] 1
        # [ ][ ][x][ ][ ][ ][ ] 2
        # [ ][ ][ ][x][ ][ ][ ] 3
        # [ ][ ][ ][ ][ ][ ][ ] 4
        # [ ][ ][ ][ ][ ][ ][ ] 5

        row = 0
        while row < self.game.rows - 3:
            column = 0
            while column < self.game.columns - 3:
                score = self.score_position(row, column, 1, 1)
                if score == self.game.score:
                    return self.game.score
                if score == -self.game.score:
                    return -self.game.score
                diagonal_points1 += score
                column += 1
            row += 1

        # Diagonal points 2 (right-bottom)
        #
        # Possible situation
        #  0  1  2  3  4  5  6
        # [ ][ ][ ][x][ ][ ][ ] 0
        # [ ][ ][x][ ][ ][ ][ ] 1
        # [ ][x][ ][ ][ ][ ][ ] 2
        # [x][ ][ ][ ][ ][ ][ ] 3
        # [ ][ ][ ][ ][ ][ ][ ] 4
        # [ ][ ][ ][ ][ ][ ][ ] 5

        row = 3
        while row < self.game.rows:
            column = 0
            while column < self.game.columns - 3:
                score = self.score_position(row, column, -1, +1)
                if score == self.game.score:
                    return self.game.score
                if score == -self.game.score:
                    return -self.game.score
                diagonal_points2 += score
                column += 1
            row += 1

        points = horizontal_points + vertical_points + diagonal_points1 + diagonal_points2
        return points

    # Determines if board is full.
    #
    # @return {boolean}

    def is_full(self):
        i = 0
        while i < self.game.columns:
            if self.field[0][i] is None:
                return False
            i += 1
        return True

    # Returns a copy of our board.
    #
    # @return {Board}

    def copy(self):
        new_board = copy.deepcopy(self)
        return new_board

File: cpp/Board.cpp
#pragma once
#include "stdafx.h"
#include "Board.h"
#include "Game.h"

Board::Board( void ) {};

Board::Board(Game* game_arg, vector< vector<int8_t>> field_arg, int8_t player_arg) : game(game_arg), field(field_arg), player(player_arg) {};

Board::~Board() {};

bool Board::isFinished(int depth, int score) {
	if ((depth == 0) | (score == this->game->score) | (score == -(this->game->score)) | this->isFull())
	{ 
		return true;
	}
	else
	{
		return false;
	}
}

bool Board::place(int column) {
	// Check if column valid
	// 1. not empty 2. not exceeding the board size
	if ((this->field[0][column] == 0) & (column >= 0) & (column < this->game->columns))
	{
		// Bottom to top
		for (int i = (this->game->rows - 1); i >= 0; i--)
		{
			if (this->field[i][column] == 0)
			{
				// Set current player coin
				this->field[i][column] = this->player;
				break;
				//# Break from loop after inserting
			}
		}
		this->player = this->game->switchRound(this->player);
		return true;
	}
	else
	{
		return false;
	}
}

int Board::scorePosition(int row, int column, int delta_y, int delta_x)
{
	int human_points = 0;
	int computer_points = 0;

	// Determine score through amount of available chips
	for (int i = 0; i < 4; i++)
	{
		if (this->field[row][column] == 1)
		{
			// Add for each human chip
			human_points++;
		}
		else if (this->field[row][column] == 2)
		{
			// Add for each computer chip
			computer_points++;
		}
		// Moving through our board
		row += delta_y;
		column += delta_x;
	}
	// Marking winning / returning score
	if (human_points == 4)
	{
		// Computer won(100000)
		return -(this->game->score);
	}
	else if (computer_points == 4)
	{
		// Human won (-100000)
		return (this->game->score);
	}
	else
	{
		// Return normal points
		return computer_points;
	}
}

int Board::score()
{

	int vertical_points = 0;
	int	horizontal_points = 0;
	int	diagonal_points1 = 0;
	int	diagonal_points2 = 0;

	// Board-size: 7x6 (height x width)
	// Array indices begin with 0
	// => e.g. height: 0, 1, 2, 3, 4, 5

	// Vertical points
	// Check each column for vertical score
	//
	// Possible situations
	//  0  1  2  3  4  5  6
	// [x][ ][ ][ ][ ][ ][ ] 0
	// [x][x][ ][ ][ ][ ][ ] 1
	// [x][x][x][ ][ ][ ][ ] 2
	// [x][x][x][ ][ ][ ][ ] 3
	// [ ][x][x][ ][ ][ ][ ] 4
	// [ ][ ][x][ ][ ][ ][ ] 5


	for (int row = 0; row < (this->game->rows - 3); row++)
	{
		//Für jede Column überprüfen
		for (int column = 0; column < (this->game->columns); column++)
		{
			// Die Column bewerten und zu den Punkten hinzufügen
			int score = this->scorePosition(row, column, 1, 0);
			if (score == this->game->score)
			{
				return this->game->score;
			}
			else if (score == -(this->game->score))
			{
				return -(this->game->score);
			}
			vertical_points += score;
		}
	}

	// Horizontal points
	// Check each row's score
	//
	// Possible situations
	//  0  1  2  3  4  5  6
	// [x][x][x][x][ ][ ][ ] 0
	// [ ][x][x][x][x][ ][ ] 1
	// [ ][ ][x][x][x][x][ ] 2
	// [ ][ ][ ][x][x][x][x] 3
	// [ ][ ][ ][ ][ ][ ][ ] 4
	// [ ][ ][ ][ ][ ][ ][ ] 5

	for (int row = 0; row < (this->game->rows); row++)
	{
		for (int column = 0; column < (this->game->columns - 3); column++)
		{
			int score = this->scorePosition(row, column, 0, 1);
			if (score == this->game->score)
			{
				return this->game->score;
			}
			else if (score == -(this->game->score))
			{
				return -(this->game->score);
			}
			horizontal_points += score;
		}
	}

	// Diagonal points 1 (left-bottom)
	//
	// Possible situation
	//  0  1  2  3  4  5  6
	// [x][ ][ ][ ][ ][ ][ ] 0
	// [ ][x][ ][ ][ ][ ][ ] 1
	// [ ][ ][x][ ][ ][ ][ ] 2
	// [ ][ ][ ][x][ ][ ][ ] 3
	// [ ][ ][ ][ ][ ][ ][ ] 4
	// [ ][ ][ ][ ][ ][ ][ ] 5

	for (int row = 0; row < (this->game->rows - 3); row++)
	{
		for (int column = 0; column < (this->game->columns - 3); column++)
		{
			int score = this->scorePosition(row, column, 1, 1);
			if (score == this->game->score)
			{
				return this->game->score;
			}
			else if (score == -(this->game->score))
			{
				return -(this->game->score);
			}
			diagonal_points1 += score;
		}
	}

	// Diagonal points 2 (right-bottom)
	//
	// Possible situation
	//  0  1  2  3  4  5  6
	// [ ][ ][ ][x][ ][ ][ ] 0
	// [ ][ ][x][ ][ ][ ][ ] 1
	// [ ][x][ ][ ][ ][ ][ ] 2
	// [x][ ][ ][ ][ ][ ][ ] 3
	// [ ][ ][ ][ ][ ][ ][ ] 4
	// [ ][ ][ ][ ][ ][ ][ ] 5

	for (int row = 3; row < (this->game->rows); row++)
	{
		for (int column = 0; column < (this->game->columns - 3); column++)
		{
			int score = this->scorePosition(row, column, -1, 1);
			if (score == this->game->score)
			{
				return this->game->score;
			}
			else if (score == -(this->game->score))
			{
				return -(this->game->score);
			}
			diagonal_points2 += score;
		}
	}

	int points = horizontal_points + vertical_points + diagonal_points1 + diagonal_points2;
	return points;
}

bool Board::isFull()
{
	for (int i = 0; i < this->game->columns; i++)
	{
		if (this->field[0][i] == 0)
		{
			return false;
		}
	}
	return true;
};

Board Board::copy()
{
	Board* board = new Board(*this);
	return *board;
};
